### ISSUE TITLE
100 NPC cause player to be kicked for airbreak very easily

### ISSUE THREAD

**Issue by rt-2** [NexiusTailer/Nex-AC#29]
Hi,
In my server I have a zombie script with many NPCs,
The AC started kicking players sometimes when they were hunting zombies.
I did my investigation, and I found that with a lot of NPCs near (from FCNPC), there was a very high chance of false positive for AirBreak.

I investigated more and the combination of around 100 NPC from FCNPC with MapAndreas, YSI and YSF is causing false positive.
I made this gamemode to help test, I tried to remove everything I could. I am not sure but I think if you remove one of the include it will not happen, but I am not sure, I tested for a long time already and I came up with this:
```
//
// Nex-Ac Test Script #1 (100 NPC cause player to be kicked for airbreak very easily)
//      by: rt-2
//      date: 2017-04-06
//
//This script will kick you for airbreak(#000) if
//	 you sprint for some time when all NPC are spawned
//

// Main Libraries
#include <a_samp>
#include <crashdetect>

//Anticheat
#include <nex-ac>

//NPCs
#include <mapandreas>
#include <FCNPC>

#include <YSI\y_master>
#include <YSF>

new npc_nb = 0;

forward CreateMyNPC();
public CreateMyNPC()
{
	if(npc_nb < 120)
	{
		new name1[16];
		format(name1, 16, "NPC_%i", npc_nb++);
		FCNPC_Create(name1);
	}
	return 1;
}

public OnGameModeInit()
{
	MapAndreas_Init(MAP_ANDREAS_MODE_FULL);
	FCNPC_InitMapAndreas(MapAndreas_GetAddress());

	SetGameModeText("Test Script");
	AddPlayerClass(random(100), 1958.3783, 1343.1572, 15.3746, 269.1425, 0, 0, 0, 0, 0, 0);
	
	
	SetTimer("CreateMyNPC", 7000, true);
	return 1;
}

public OnPlayerSpawn(playerid)
{
	return 1;
}
public FCNPC_OnCreate(npcid)
{
	FCNPC_Spawn(npcid, random(100), 1958.3783, 1344.1572, 15.3746);
}

main()
{
	print("\n----------------------------------");
	printf(" 'Nex-Ac Test Script' by rt-2 ");
	print("----------------------------------\n");
}
```

In this gamemode, when all the NPC will be spawned, you wont be able to run around for more than a couple minutes, sometime seconds, without being kicked for AirBreak (code #00). Also I think bunny hoping will help to get kicked but I'm not sure.

Here is link to the includes needed, they are all updated in my test:
https://github.com/ziggi/FCNPC
http://forum.sa-mp.com/showpost.php?p=3130004&postcount=153
https://github.com/Misiur/YSI-Includes

I know the problem might be in one of those includes, and it is probably, but you are the only one who will be able to tell, all I know is that with this combination (that is needed for my server script), we will get kicked for AirBreak when near a large number of NPCs.

I tested this with 2 server, my main one on the internet, and also I tested from my own computer and connected to myself.
Sometime it can take 5-10minutes max to get kicked, sometimes 5-10 seconds, but you will always get kicked at some point if you run around in short streaming distance of the NPCs for an extended period of time.

Thank you, 
rt-2

**Comment by rt-2** [comment]
I reduced the number to 50 NPCs on my server, it almost never happens now but it happened to me once this morning.
Although, the more NPCs there is, the more chance you have to be kicked for Airbreak.
Is there any advancements on this issue? Have you been able to test it?

Thank you very much,
rt-2

**Comment by rt-2** [comment]
I edited the first post, YSF doesn't "need" to be included.

**Comment by NexiusTailer** [comment]
Hi, I apologize for such a long response.
Yes, the problem is also reproduced with me, and I have some ideas for solving it (reducing some settings, most likely [these values](https://github.com/NexiusTailer/Nex-AC/blob/master/src/v1.9/v1.9.23/r2/nex-ac.inc#L4333)). I'll let you know as soon as I can fix it.

Many thanks!

**Comment by rt-2** [comment]
Thank you very much.
Good luck, tell me if I can help.
Thanks,
rt-2

**Comment by NexiusTailer** [comment]
Yes, it seems that the problem has disappeared after some changes on this line:
https://github.com/NexiusTailer/Nex-AC/blob/master/src/v1.9/v1.9.23/r2/nex-ac.inc#L4333

I changed the multiplier "30.0" to "26.0". At the same time, I don't observe any serious deterioration in the detection of a real airbreak. I think it's worth check this out some time...

**Comment by rt-2** [comment]
I still have problems on my side,
I have put the npc number to 80 and, when I run uphill or sprint, I get kicked almost all the time when near the NPCs.
Change ``if(npc_nb < 120)`` to ``if(npc_nb < 80)`` in ``CreateMyNPC()``.
The "speed" in the debug prints have been as low as 20.0-21.0, and even 16.0.
Thank you, 
rt-2

**Comment by rt-2** [comment]
Maybe the problem comes from one of the include needed to reproduce the problem (YSI and/or YSF)?
But I don't know the internal functioning of this mechanism.
Thank you,
rt-2

**Comment by NexiusTailer** [comment]
Hmm, I checked this at about 80 NPCs with the multiplier set to "26.0" as I wrote earlier, and this time I wasn't able to get a false detection not once. Also it does not seem that problem depend on any other ones like YSI or YSF.

**Comment by rt-2** [comment]
Personnaly I compiled this:
```
//
// Nex-Ac Test Script #1 (100 NPC cause player to be kicked for airbreak very easily)
//      by: rt-2
//      date: 2017-04-06
//
//This script will kick you for airbreak(#000) if
//	 you sprint for some time when all NPC are spawned
//

// Main Libraries
#include <a_samp>
#include <crashdetect>

//Anticheat
#include <nex-ac>

//NPCs
#include <mapandreas>
#include <FCNPC>

#include <YSI\y_master>
#include <YSF>

new npc_nb = 0;

forward CreateMyNPC();
public CreateMyNPC()
{
	if(npc_nb < 80)
	{
		new name1[16];
		format(name1, 16, "NPC_%i", npc_nb++);
		FCNPC_Create(name1);
	}
	return 1;
}

public OnGameModeInit()
{
	MapAndreas_Init(MAP_ANDREAS_MODE_FULL);
	FCNPC_InitMapAndreas(MapAndreas_GetAddress());

	SetGameModeText("Test Script");
	AddPlayerClass(random(100), 1958.3783, 1343.1572, 15.3746, 269.1425, 0, 0, 0, 0, 0, 0);


	SetTimer("CreateMyNPC", 7000, true);
	return 1;
}

public OnPlayerSpawn(playerid)
{
	return 1;
}
public FCNPC_OnCreate(npcid)
{
	FCNPC_Spawn(npcid, random(100), 1958.3783, 1344.1572, 15.3746);
}

main()
{
	print("\n----------------------------------");
	printf(" 'Nex-Ac Test Script' by rt-2 ");
	print("----------------------------------\n");
}
```
And I got kicked very easily, especially when my Z changed.
My sweet spot to make it happen on this script is on the escalator, going up or down.
Also on my server it seemed to happen more in hilly regions.
I got this in the console:
```
[21:04:12] [connection] 127.0.0.1:50324 requests connection cookie.
[21:04:13] [connection] incoming connection: 127.0.0.1:50324 id: 0
[21:04:13] [join] Yanick_Coulombe has joined the server (0:127.0.0.1)
[21:04:25] [Nex-AC debug] Speed: 19.0, distance: 0.824346
[21:04:25] [Nex-AC] Suspicious player, ID 0. Reason code: 000
[21:04:25] [part] Yanick_Coulombe has left the server (0:2)
[21:05:32] [connection] 127.0.0.1:56457 requests connection cookie.
[21:05:33] [connection] incoming connection: 127.0.0.1:56457 id: 0
[21:05:33] [join] Yanick_Coulombe has joined the server (0:127.0.0.1)
[21:05:48] [Nex-AC debug] Speed: 20.0, distance: 0.788956
[21:05:48] [Nex-AC] Suspicious player, ID 0. Reason code: 000
[21:05:48] [part] Yanick_Coulombe has left the server (0:2)
[21:06:26] [connection] 127.0.0.1:51166 requests connection cookie.
[21:06:27] [connection] incoming connection: 127.0.0.1:51166 id: 0
[21:06:28] [join] Yanick_Coulombe has joined the server (0:127.0.0.1)
[21:06:35] [Nex-AC debug] Speed: 21.0, distance: 0.847419
[21:06:35] [Nex-AC] Suspicious player, ID 0. Reason code: 000
[21:06:35] [part] Yanick_Coulombe has left the server (0:2)
```
Would it be possible that network quality and/or server performance may affect the inputs of the anti cheat?
Thank you,
rt-2

**Comment by rt-2** [comment]
Any clue on what could be causing this? the source of the problem?
Thank you,
rt-2

**Comment by NexiusTailer** [comment]
When I test it, I noticed that the amount of traffic of the client with the server is '''significantly''' increased with the increase of NPC in the stream, which can be observed in game by pressing the F5 button. I guess that processing a large amount of information increases the chance of lagging some data, so this may well be the reason.

As a solution, you can increase the [number of warnings for anti-airbreak](https://github.com/NexiusTailer/Nex-AC/blob/master/src/v1.9/v1.9.24/r2/nex-ac.inc#L84), after which the anticheat will punish the player... or you can reduce the multiplier again, which I changed in the latest version, but this can affect the quality of the detection of airbreak at short distances

**Comment by rt-2** [comment]
I would want to open a issue on the FCNPC github, but before I do, do you have any idea on how a high network traffic can affect your calculation?
And would there be another way of verifying that the information we receive is not outdated? like with the TickCount, maybe if see that the last time we successfully tested a position was at sa certain tick, and now we have way more ticks and maybe the difference could help calculate what distance could be allowed between those 2 times, instead of relying on a constant PlayerUpdate?
I'm not sure if this makes sense because I'm not too familiar with the internal functioning of this AC, but if you check what time and the difference of time between tests, would there be a way to overcome those network slowdowns?
Thank you,
rt-2

**Comment by NexiusTailer** [comment]
I think this is definitely not the problem of the FCNPC plugin, because It only gives some functionality for working with bots, and the problem is in some delays in those places where you need to synchronize a lot of data (for example, a huge number of players/nps in the stream).
I think it's really better to try changing the maximum value of the airbreak warnings before changing the formula again (that is responsible for detecting the cheat).

**Comment by rt-2** [comment]
I think we should find a way to compensate for network lags, at least know when it is happening, and compensate in the value, for how late we received the packet, or maybe the packets are not received in the right order...
Do you think it's possible?
It would mean the anti-cheat would be more stable for all types of detection.

**Comment by NexiusTailer** [comment]
Yes it's possible but I'm not sure that this will fix everything... I think it's most likely impossible to know exactly the time of the delay between the sending of new speed and new position, only the total delay, which is expressed in ping.
Perhaps there is a variant of using NetStats_* functions (for example NetStats_MessagesRecvPerSecond) to find out the amount of sended and received data for a certain time (because with a large number of other players in the stream this data will increase), but these are just my guesses and I don't think that this an effective solution of the problem.

**Comment by rt-2** [comment]
Alright,
We'll keep this opened,
We'll try to find a solution, I'll try to do some research/tests with NetStats functions, maybe with YSF or SKY to see their networks/RPC functions,
Thank you,
rt-2

**Comment by rt-2** [comment]
I have made researches, I tested some things.
I have a question, the timer 'ac_timer', it's called each 1000ms, if let say your anticheat detect a cheat, but this timer has been called last time 1050-1300ms away,, may that indicate a cpu overload that would result in a false flag maybe?
Something like this: https://github.com/NexiusTailer/Nex-AC/pull/42
Thank you,
rt-2

**Comment by NexiusTailer** [comment]
@rt-2 Good idea. But could you check how [this](GetServerTickRate) (GetServerTickRate) function will behave under such overload? (it seems that it is intended for this cases and it would be possible to do some checks with it)

**Comment by rt-2** [comment]
Just now me and a buddy were hunting 44 zombies(NPC), we both got kicked once, me with a tick difference of 55422 and him 46453. I'm testing with the server tick rate, as of now, 130-170 seems to be a normal value.

**Comment by rt-2** [comment]
I have tested on my server the "server rate", it seems to always stay between 120 and 190 and stays well between those limits when comes a 'faulty' kick. By faulty kicks I mean kicks with a tick difference of more than 1030ms after the last ac_timer, here is some values I collected (note that at first, I was not collecting server rates):
```
Tick Diff: 1033/1000
Tick Diff: 1098/1000
Tick Diff: 1049/1000
Tick Diff: 46453/1000 *
Tick Diff: 55422/1000 *
Tick Diff: 1076/1000
Tick Diff: 1040/1000 with server rate:168
Tick Diff: 1035/1000 with server rate:162
Tick Diff: 1117/1000 with server rate:171
Tick Diff: 1106/1000 with server rate:166
Tick Diff: 1037/1000 with server rate:156
Tick Diff: 1183/1000 with server rate:157
Tick Diff: 1156/1000 with server rate:174
Tick Diff: 1142/1000 with server rate:175
```
* Those 2 we're happening because of the NPCs, other result I think are mistakes because of CPU overload.
Thank you,
Yanick

**Comment by NexiusTailer** [comment]
So now I changed values '26.0' to '20.0' on this lines:
https://github.com/NexiusTailer/Nex-AC/blob/master/src/v1.9/v1.9.27/r2/nex-ac.inc#L4365-L4366

And as far as I've tested it for a long time, I didn't find any more false detections with a lot of NPCs. Also try to change this value for yourself as a temporary solution, because it will be implemented in v1.9.28.
Let me know if you find something or if it does not help you.
